### PR TITLE
ProjectKorra: Rebirth

### PR DESCRIPTION
**Pull request by StrangeOne101**
# Changes
These are general changes users/server owners will see and notice.
- Added hex colors for each subelement
- Added a `/b cooldown` command to view/set/reset cooldowns for a player
- Made the config generate blocks based on the current server version
- Added better support for 1.16
- The \[ProjectKorra\] prefix in front of commands is completely configurable
- Commands can now affect offline players
- The combinations used to trigger combos is now configurable
- Reimplemented day/night factors
- Better EarthArmor colors
- Removed autoannouncer (unneeded, no one wants it)

# Developer Changes
These are changes that benefit developers of side plugins and addon abilities

- Added TempFallingBlock class
- Added PlayerSwingEvent for addon developers (better than the interact or animation event as it contains all the checks PK does to make sure users aren't interacting with anything)
- Added OfflineBendingPlayer class for offline player data
- Cleaned up GeneralMethods
- Made addons who don't define permissions have permissions defined for them instead
- Added BendingRegionProtection class so 3rd parties can add bending support
- Added MultiSubElement class - a class for subelements that have more than 1 parent element (e.g. mudbending for water and earth)

# Fixes
- Fixed the server crashing when an ability's range is 0 (#1212)
- Fixed addons registering their listeners twice on /b reload
- Fixed addon ability jars being locked so they couldn't be deleted while the server was on (excluding Linux, which doesn't care if it SHOULD delete it or not)
- Fixed presets halting main thread when being created (caused a lot of lag)
- Fixed presets halting main thread when being deleted
- Fixed preset tabbing not working for binding and deleting
- Fixed vanished players being visible in tab completion
- Fixes BendingBoard IllegalStateException thrown when team is unregistered twice
- Fixed Extraction not working on 1.16 servers
- Fixed cooldowns halting the main thread when saving (and causing a lot of lag)
- Fixed ALL cooldowns being saved to the database
- Fixed bending boards showing up in disabled worlds when you log in
- Fixed bending toggle reminder not being translatable
- Fixed "Proper Usage: xxx" in commands being untranslatable
- Fixes addon abilities that do not define their own permissions being unusable until the permission is set by the server owner
- Fixed subs not reloading their colors from config on pk reload
- Fixed branding messages not supporting hex colors
- Fixed bending toggle reminder not being translatable
- Fixed untranslated line for failing to bind when you have a multiability bound
- Fixed temp snow blocks not preventing fall damage in HydroSink (the water passive)
- Fixed memory leak where BendingPlayer objects are never unloaded from memory
- Fixed toggled bending not persisting over relogs
- Fixed `/b choose avatar` not giving all 4 elements but instead giving the internal "Avatar" element that is useless
- Fixed Night and Day factors not existing for 90% of abilities
- Fixed the blue fire factor not being used for a lot of abilities
- Fixed earth abilities not using the Metal power factor for damage when using a metal source
- Fixed combos being triggered from block placing
- #1192 
- Fixed the copy command able to copy while using a multi ability

### CODE REVIEW

**Review by Lumos-Tyson — COMMENTED** [review]
Hey Toby, it's obvious you spent a ton of time on this, are there are a bunch of great changes. I love the ideas behind the modifier changes, using fast math, and cleaning up some extraneous stuff throughout the different classes.

I left behind just a few comments. Thanks!

File: src/com/projectkorra/projectkorra/Element.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/Element.java:129]: This is because subelements having individual colors is very new. If a color is not defined for them, it will then fall back to using the default subelement color defined in the parent element like before

File: src/com/projectkorra/projectkorra/ProjectKorra.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/ProjectKorra.java:61]: Left in accidentally. Thanks for spotting

File: src/com/projectkorra/projectkorra/ability/CombustionAbility.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/ability/CombustionAbility.java:24]: Lol true. Copied and pasted it

File: src/com/projectkorra/projectkorra/earthbending/EarthArmor.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/earthbending/EarthArmor.java:438]: Yeah, I made a script to extract the average color from textures just for this lol

File: src/com/projectkorra/projectkorra/earthbending/EarthBlast.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/earthbending/EarthBlast.java:124]: Correct. In the config, we have always had a metal power factor where using metal sources would do more damage. But it has been left to rot like how day/night buffs were

File: src/com/projectkorra/projectkorra/earthbending/metal/Extraction.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/earthbending/metal/Extraction.java:83]: I haven't touched the amounts, it was already like that. Perhaps ask @Aztlon

**Comment by StrangeOne101** [comment]
Thanks for the review. You picked up a few things I missed

File: src/com/projectkorra/projectkorra/ability/LightningAbility.java
Comment (Varhagna) [inline src/com/projectkorra/projectkorra/ability/LightningAbility.java:30]: See previous comment on Combustion & Blue Fire

File: src/com/projectkorra/projectkorra/airbending/AirBurst.java
Comment (Varhagna) [inline src/com/projectkorra/projectkorra/airbending/AirBurst.java:124]: See previous comment on FastMath

File: src/com/projectkorra/projectkorra/GeneralMethods.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:843]: Please see https://www.element84.com/blog/improving-java-math-performance-with-jafama for speed comparasions.

Yes, I know that in the grand scheme of things, micro optimizations won't help a lot. But that isn't the point. This is technically faster, and if it's available, there is no reason to not use these. The inaccuracy of FastMath compared to Math will only really be seen in the 5th decimal place or further. Meaning that for what we use, it is impossible to see a difference (as you'd need something bigger than 10,000 blocks for it to be a block out of where it should be)

File: src/com/projectkorra/projectkorra/GeneralMethods.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:1710]: This was simply a find and replace. I wasn't optimizing anything further. It can be done in a future update if needed imo

File: src/com/projectkorra/projectkorra/PKListener.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/PKListener.java:1423]: Why? It's easier to read over multiple lines when you don't have to scroll needlessly

File: src/com/projectkorra/projectkorra/ability/CombustionAbility.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/ability/CombustionAbility.java:25]: Well the blue fire factor doesn't even work properly right now. This is still better than not doing it at all. But I think most people will agree with me that combustion and lightning shouldn't get buffed by it

File: src/com/projectkorra/projectkorra/GeneralMethods.java
Comment (Simplicitee) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:843]: I'm not sure what all he did, but Vahagn did some benchmarks on FastMath compared to Math and it seemed to be slower for his tests.

Comment (Varhagna) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:843]: Yes I looked into that link provided, however I decided to run my own benchmarks reflecting the typical PK use case as it bothered my intuition. My own benchmarks contradict those that were done in that web blog. 
```

public class MathBenchmarkDriver {

	public static void main(String[] args) {
		long a = 0;
		long b = 0;
		
		int n = 360;

		long startTime_1 = 0;
		long endTime_1 = 0;
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_1 = System.nanoTime();
			FastMath.sin(in);
			endTime_1 = System.nanoTime();
			a += (endTime_1 - startTime_1);
		}

		
		long startTime_2 = 0;
		long endTime_2 = 0;
		
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_2 = System.nanoTime();
			Math.sin(in);
			endTime_2 = System.nanoTime();
			b += (endTime_2 - startTime_2);
		}
		
		double avg_a = a / n;
		double avg_b = b / n;
		
		System.out.println("Sin Math: " + avg_b);
		System.out.println("Sin FastMath: " + avg_a);
		System.out.println("Sin Diff: " + (avg_b - avg_a));
		
		a = 0;
		b = 0;
		
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_1 = System.nanoTime();
			FastMath.cos(in);
			endTime_1 = System.nanoTime();
			a += (endTime_1 - startTime_1);
		}
		
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_2 = System.nanoTime();
			Math.cos(in);
			endTime_2 = System.nanoTime();
			b += (endTime_2 - startTime_2);
		}
		
		avg_a = a / n;
		avg_b = b / n;
		
		System.out.println("Cos Math: " + avg_b);
		System.out.println("Cos FastMath: " + avg_a);
		System.out.println("Cos Diff: " + (avg_b - avg_a));
		
		a = 0;
		b = 0;

		for(int i = 0; i < n; i++) {
			double in = Math.random() * n;
			startTime_1 = System.nanoTime();
			FastMath.sqrt(in);
			endTime_1 = System.nanoTime();
			a += (endTime_1 - startTime_1);
		}
		
		for(int i = 0; i < n; i++) {
			double in = Math.random() * n;
			startTime_2 = System.nanoTime();
			Math.sqrt(in);
			endTime_2 = System.nanoTime();
			b += (endTime_2 - startTime_2);
		}
		
		avg_a = a / n;
		avg_b = b / n;
		
		System.out.println("Sqrt Math: " + avg_b);
		System.out.println("Sqrt FastMath: " + avg_a);
		System.out.println("Sin Diff: " + (avg_b - avg_a));
		
	}
}

```

For n = 360 (typical PK usecase)

image
image


For n = 500000

image
image

For n = 1000000

image
image

I dont think the performance hit for n < 500,000 is justifies using it.

Comment (Varhagna) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:843]: I doubted the benchmarks on that website when I did my own preliminary search since it violated my intuition.

After running benchmarks using this
```

public class MathBenchmarkDriver {

	public static void main(String[] args) {
		long a = 0;
		long b = 0;
		
		int n = 360;
		
		for(int i = 0; i < 1000000; i++) {
			double in = Math.toRadians(i % 360);
			FastMath.sin(in);
			Math.sin(in);
			FastMath.cos(in);
			Math.cos(in);
			FastMath.sqrt(in);
			Math.sqrt(in);
		}
		
		long startTime_1 = 0;
		long endTime_1 = 0;
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_1 = System.nanoTime();
			FastMath.sin(in);
			endTime_1 = System.nanoTime();
			a += (endTime_1 - startTime_1);
		}

		
		long startTime_2 = 0;
		long endTime_2 = 0;
		
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_2 = System.nanoTime();
			Math.sin(in);
			endTime_2 = System.nanoTime();
			b += (endTime_2 - startTime_2);
		}
		
		double avg_a = a / n;
		double avg_b = b / n;
		
		System.out.println("Sin Math: " + avg_b);
		System.out.println("Sin FastMath: " + avg_a);
		System.out.println("Sin Diff: " + (avg_b - avg_a));
		
		a = 0;
		b = 0;
		
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_1 = System.nanoTime();
			FastMath.cos(in);
			endTime_1 = System.nanoTime();
			a += (endTime_1 - startTime_1);
		}
		
		for(int i = 0; i < n; i++) {
			double in = Math.toRadians(i % 360);
			startTime_2 = System.nanoTime();
			Math.cos(in);
			endTime_2 = System.nanoTime();
			b += (endTime_2 - startTime_2);
		}
		
		avg_a = a / n;
		avg_b = b / n;
		
		System.out.println("Cos Math: " + avg_b);
		System.out.println("Cos FastMath: " + avg_a);
		System.out.println("Cos Diff: " + (avg_b - avg_a));
		
		a = 0;
		b = 0;

		for(int i = 0; i < n; i++) {
			double in = Math.random() * n;
			startTime_1 = System.nanoTime();
			FastMath.sqrt(in);
			endTime_1 = System.nanoTime();
			a += (endTime_1 - startTime_1);
		}
		
		for(int i = 0; i < n; i++) {
			double in = Math.random() * n;
			startTime_2 = System.nanoTime();
			Math.sqrt(in);
			endTime_2 = System.nanoTime();
			b += (endTime_2 - startTime_2);
		}
		
		avg_a = a / n;
		avg_b = b / n;
		
		System.out.println("Sqrt Math: " + avg_b);
		System.out.println("Sqrt FastMath: " + avg_a);
		System.out.println("Sin Diff: " + (avg_b - avg_a));
		
	}
}
```

For n = 500000
image
image

We see variable but always nearly the same performance between the two.

For n =  360 (typical PK use case)

image
image

We see consistently that Math.sin performs better than FastMath.cos. FastMath.cos and Math.cos have variable but equal performance. and Math.sqrt vs FastMath.sqrt is by default going to delegate to Math.sqrt, therefore its always slightly faster due to the additional function call overhead.

With this in mind, I don't see any reason to use this library, because the performance is marginally better if at all, and because its entirely written in Java unlike Math library native code, cannot take advantage of any low level hardware optimizations that hosts often implement to boost performance.

File: src/com/projectkorra/projectkorra/GeneralMethods.java
Comment (Varhagna) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:1710]: Gotcha

File: src/com/projectkorra/projectkorra/PKListener.java
Comment (Varhagna) [reply src/com/projectkorra/projectkorra/PKListener.java:1423]: To keep the code standard.

As far as I am aware, the standard has always been that its always on one line regardless of length. 
@ChristopherWMM can correct me if I am wrong on this.

Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/PKListener.java:1423]: I don't think this is standard, and even if it was, I don't think there is much point enforcing a (bad) standard on a codebase that's going to be thrown away anyway

Comment (Aztlon) [reply src/com/projectkorra/projectkorra/PKListener.java:1423]: There is no standard bahaha just do what looks better

File: src/com/projectkorra/projectkorra/GeneralMethods.java
Comment (Aztlon) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:843]: so true king don't use FastMath

File: src/com/projectkorra/projectkorra/earthbending/metal/Extraction.java
Comment (Aztlon) [reply src/com/projectkorra/projectkorra/earthbending/metal/Extraction.java]: oh so this is why i kept getting emails from this pr. it's multiplied because it drops gold nuggets instead of ingots but i kinda forgot why it does that.

File: src/com/projectkorra/projectkorra/GeneralMethods.java
Comment (StrangeOne101) [reply src/com/projectkorra/projectkorra/GeneralMethods.java:2035]: Confirmed the result is the same from both of these functions. Will switch to this as it is more simple

File: src/com/projectkorra/projectkorra/PKListener.java
Comment (ChristopherWMM) [reply src/com/projectkorra/projectkorra/PKListener.java:1423]: This is as superficial as it gets and should not be a reason to hold this up. Heck, an IDE probably did this automatically.

File: src/com/projectkorra/projectkorra/ProjectKorra.java
Comment (ChristopherWMM) [reply src/com/projectkorra/projectkorra/ProjectKorra.java]: @StrangeOne101 this should be good to resolve, correct?

File: pom.xml
Comment (ChristopherWMM) [reply pom.xml:6]: Absolutely! Especially with our goal to be releasing smaller, singularly focused, builds. Moving forward, bumping the minor version should only be reserved for quick-fixes/patches.

**Review by Varhagna — APPROVED** [review]
My name is Vahagn, and I approve this PR.

😄

**Review by Lumos-Tyson — APPROVED** [review]
Great work, Strangeone!

**Comment by PrideInt** [comment]
just want to say i approve before chris merges :)

**Pull request merged**
